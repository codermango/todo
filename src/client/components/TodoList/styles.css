.todoList {
  margin: 10px;
}
