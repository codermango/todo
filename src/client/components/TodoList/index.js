import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import TodoItem from '../TodoItem';
import { updateAll } from '../../actions';

import styles from './styles.css';

class TodoList extends Component {

  static propTypes = {
    todos: PropTypes.array.isRequired,
    updateAll: PropTypes.func.isRequired
  }

  constructor(props) {
    super(props);
    this.draggedItem = null;
  }

  handleDragStart = (idx) => (e) => {
    const { todos = [] } = this.props;
    this.draggedItem = todos[idx];
  }

  handleDragOver = (idx) => (e) => {
    const { todos = [], updateAll } = this.props;
    const draggedOverItem = todos[idx];
    if (this.draggedItem.id === draggedOverItem.id) {
      return;
    }

    let newTodos = todos.filter(todo => todo.id !== this.draggedItem.id);
    newTodos.splice(idx, 0, this.draggedItem);
    updateAll({ todos: newTodos });
  }

  render() {
    const { todos = [] } = this.props;

    return (
      <div className={styles.todoList} data-test-id="todo-list">
        {todos.map((todo, idx) => (
          <TodoItem
            key={todo.id}
            data={todo}
            draggable={true}
            onDragStart={this.handleDragStart(idx)}
            onDragOver={this.handleDragOver(idx)}
          />
        ))}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  todos: state.todos
})

const mapDispatchToProps = dispatch => ({
  updateAll: ({ todos }) => dispatch(updateAll({ todos }))
});

export default connect(mapStateToProps, mapDispatchToProps)(TodoList);
