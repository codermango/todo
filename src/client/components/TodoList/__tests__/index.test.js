import React from 'react';
import { shallow } from 'enzyme';
import TodoList from '../index';


describe('<TodoList />', () => {

  const props = {
    todos: [
      { id: 'id1', text: 'text1', done: false },
      { id: 'id2', text: 'text2', done: false },
      { id: 'id3', text: 'text3', done: false }
    ],
    updateAll: jest.fn()
  };

  it('renders correctly', () => {
    const wrapper = shallow(<TodoList {...props} />);
    const instance = wrapper.instance();
    expect(instance).toMatchSnapshot();
  });

  it('handleDragStart works correctly', () => {
    const wrapper = shallow(<TodoList {...props} />);
    const instance = wrapper.instance();
    instance.handleDragStart(0)(null);
    expect(instance.draggedItem).toEqual(props.todos[0]);
  });

  it('handleDragOver works correctly', () => {
    const wrapper = shallow(<TodoList {...props} />);
    const instance = wrapper.instance();
    instance.handleDragStart(0)(null);
    instance.handleDragOver(1)(null);
    expect(instance.props.updateAll).toHaveBeenCalledTimes(1);

    instance.props.updateAll.mockClear();
    instance.handleDragStart(0)(null);
    instance.handleDragOver(0)(null);
    expect(instance.props.updateAll).toHaveBeenCalledTimes(0);
  });
});
