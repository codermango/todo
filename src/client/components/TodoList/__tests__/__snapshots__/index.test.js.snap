// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TodoList /> renders correctly 1`] = `
TodoList {
  "context": Object {},
  "draggedItem": null,
  "handleDragOver": [Function],
  "handleDragStart": [Function],
  "props": Object {
    "todos": Array [
      Object {
        "done": false,
        "id": "id1",
        "text": "text1",
      },
      Object {
        "done": false,
        "id": "id2",
        "text": "text2",
      },
      Object {
        "done": false,
        "id": "id3",
        "text": "text3",
      },
    ],
    "updateAll": [MockFunction],
  },
  "refs": Object {},
  "setState": [Function],
  "state": null,
  "updater": Updater {
    "_callbacks": Array [],
    "_renderer": ReactShallowRenderer {
      "_context": Object {},
      "_didScheduleRenderPhaseUpdate": false,
      "_dispatcher": Object {
        "readContext": [Function],
        "useCallback": [Function],
        "useContext": [Function],
        "useDebugValue": [Function],
        "useEffect": [Function],
        "useImperativeHandle": [Function],
        "useLayoutEffect": [Function],
        "useMemo": [Function],
        "useReducer": [Function],
        "useRef": [Function],
        "useState": [Function],
      },
      "_element": <TodoList
        todos={
          Array [
            Object {
              "done": false,
              "id": "id1",
              "text": "text1",
            },
            Object {
              "done": false,
              "id": "id2",
              "text": "text2",
            },
            Object {
              "done": false,
              "id": "id3",
              "text": "text3",
            },
          ]
        }
        updateAll={[MockFunction]}
      />,
      "_firstWorkInProgressHook": null,
      "_forcedUpdate": false,
      "_instance": [Circular],
      "_isReRender": false,
      "_newState": null,
      "_numberOfReRenders": 0,
      "_renderPhaseUpdates": null,
      "_rendered": <div
        className="todoList"
        data-test-id="todo-list"
      >
        <TodoItem
          data={
            Object {
              "done": false,
              "id": "id1",
              "text": "text1",
            }
          }
          draggable={true}
          onDragOver={[Function]}
          onDragStart={[Function]}
        />
        <TodoItem
          data={
            Object {
              "done": false,
              "id": "id2",
              "text": "text2",
            }
          }
          draggable={true}
          onDragOver={[Function]}
          onDragStart={[Function]}
        />
        <TodoItem
          data={
            Object {
              "done": false,
              "id": "id3",
              "text": "text3",
            }
          }
          draggable={true}
          onDragOver={[Function]}
          onDragStart={[Function]}
        />
      </div>,
      "_rendering": false,
      "_updater": [Circular],
      "_workInProgressHook": null,
    },
  },
}
`;
